util/network-devp2p/src/discovery.rs

```diff
@@ -591,65 +658,74 @@ impl<'a> Discovery<'a> {
 		Ok(None)
 	}
 
-	fn check_expired(&mut self, time: Instant) -> HashSet<NodeId> {
-		let mut removed: HashSet<NodeId> = HashSet::new();
-		while let Some((node_id, sent_at)) = self.expiring_pings.pop_front() {
-			if time.duration_since(sent_at) <= PING_TIMEOUT {
-				self.expiring_pings.push_front((node_id, sent_at));
-				break;
+	fn check_expired(&mut self, time: Instant) {
+		let mut nodes_to_expire = Vec::new();
+		self.in_flight_pings.retain(|node_id, ping_request| {
+			if time.duration_since(ping_request.sent_at) > PING_TIMEOUT {
+				debug!(target: "discovery", "Removing expired PING request for node_id=0x{:x}", node_id);
+				nodes_to_expire.push(*node_id);
+				false
+			} else {
+				true
 			}
-			self.expire_in_flight_request(node_id, sent_at, &mut removed);
-		}
-		while let Some((node_id, sent_at)) = self.expiring_finds.pop_front() {
-			if time.duration_since(sent_at) <= FIND_NODE_TIMEOUT {
-				self.expiring_finds.push_front((node_id, sent_at));
-				break;
+		});
+		self.in_flight_find_nodes.retain(|node_id, find_node_request| {
+			if time.duration_since(find_node_request.sent_at) > FIND_NODE_TIMEOUT {
+				if !find_node_request.answered {
+					debug!(target: "discovery", "Removing expired FIND NODE request for node_id=0x{:x}", node_id);
+					nodes_to_expire.push(*node_id);
+				}
+				false
+			} else {
+				true
 			}
-			self.expire_in_flight_request(node_id, sent_at, &mut removed);
+		});
+		for node_id in nodes_to_expire {
+			self.expire_node_request(node_id);
 		}
-		removed
```

Comment:
So when are unreliable nodes removed from the stored table now?

Reply:
The idea is that they are not removed via the discovery protocol, but the actual TCP protocol has some mechanisms to remove unresponsive nodes (or ones on another chain, etc.)
The issue was for nodes that have UDP blocked and don't run with `--no-discovery`. The number of peers would slowly drop to 0 and we won't be able to recover from that.
I can run some more tests to see how it affects the sync though.

Reply:
Cool, seems reasonable.